dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ServiceConfig.java

```diff
@@ -633,6 +633,25 @@ private String findConfigedHosts(ProtocolConfig protocolConfig, List<URL> regist
      * @return
      */
     private Integer findConfigedPorts(ProtocolConfig protocolConfig, String name, Map<String, String> map) {
+        Integer portToBind = getPortFromProtocol(protocolConfig, name);
+
+        // save bind port, used as url's key later
+        map.put(Constants.BIND_PORT_KEY, String.valueOf(portToBind));
+
+        // registry port, not used as bind port by default
+        String portToRegistryStr = getValueFromConfig(protocolConfig, Constants.DUBBO_PORT_TO_REGISTRY);
+        Integer portToRegistry = parsePort(portToRegistryStr);
+        if (portToRegistry == null) {
+            portToRegistry = portToBind;
+        }
+
+        return portToRegistry;
+    }
+
+    private Integer getPortFromProtocol(ProtocolConfig protocolConfig, String name) {

```

Comment:
I don't think this is a good method name.
When the port is missed in config, it will be found in system or protocol instance or somewhere.

Reply:
I just think that current logic is just that it does duplicate work too many times. We can get available port from system properties, dubbo config properties or the config of xml. But it does not have to do the same thing in every dubbo service export process. It seems redundant

Reply:
If the port is strongly related to the protocol, can we move the find port logic to the process of generate the protocol config bean ? It can not only reduce the duplicate execution of find port logic, but also can present a more clear logic.

Reply:
I think current code is like that we have known the result of one computing problem after we solved it first time. But when we need the result, we still try to solve the problem once by once, but not to get the result directly. But the problem and the result does not change always.

Reply:
Seems like most time we just have one protocol. So it is not 'once by once' like u said.
ProtocolConfig's port is set by user. U can fix it by other way, but not to change the user config.

Reply:
'Once by once' that I said means that when every <dubbo:service> that use the same protocol export, it will execute the same logic once.

Reply:
I only have one problem. We can use serviceConfig and protocolConfig to publish the service right? If we accept your pr, the user does not set the port of protocolConfig before the release. After the release, protocolConfig has a port. How do we explain this to user that protocolConfig auto set a port?

Reply:
I don't think this problem just from the user config side. I think if the user does not set the port of the protocol, it means that he does care the port and he just want the application to run correctly. So I think the port of protocol can regards as an inner state that the user can ignore as if the application can run with no problems.

Reply:
I think common users of dubbo don’t care what class represents the <dubbo:protocol>, and they don’t have to know how dubbo run. If they don’t config a port clearly, they may or may not know dubbo will pick an available port for them. But I think it’s not important that the common users do know or do not know. They can only care about if dubbo run correctly. The inner running mechanism can be transparent to the common users.

Reply:
If the user does not set the port, it of course means that he does not care about the port. It is common sense.